File: src/cpp/SpiralPrinting.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

using namespace std;

typedef vector< vector<string> > Matrix;

void parseMatrix(Matrix & m, string s) {
    int rows, cols;
    char c;
    stringstream ss(s);
    ss >> rows;
    ss >> c;
    ss >> cols;
    ss >> c;
    m.clear();
    for (int i = 0; i < rows; i++) {
        vector<string> row;
        for (int j = 0; j < cols; j++) {
            string s;
            ss >> s;
            row.push_back(s);
        }
        m.push_back(row);
    }
}

Matrix rotateCCW(const Matrix & m) {
    int rows = m.size();
    int cols = m[0].size();
    Matrix rotated;
    for (int i = 0; i < cols; i++) {
        vector<string> row(rows, "");
        rotated.push_back(row);
    }
    for (int i = 0; i < rows; i++) {
        for (int j = cols - 1; j >= 0; j--) {
            rotated[cols - 1 - j][i] = m[i][j];
        }
    }
    return rotated;
}

void printFirstRow(const Matrix & m) {
    vector<string> row = m[0];
    for (unsigned int i = 0; i < row.size(); i++) {
        cout << row[i];
        if (i != row.size() - 1) {
            cout << ' ';
        }
    }
}

void removeFirstRow(Matrix & m) {
    m.erase(m.begin());
}

int main(int argc, char* argv[]) {
    ifstream infile(argv[1]);
    string line;
    Matrix m;
    while (getline(infile, line)) {
        parseMatrix(m, line);
        while (m.size() > 0) {
            printFirstRow(m);
            removeFirstRow(m);
            if (m.size() > 0) {
                cout << ' ';
                m = rotateCCW(m);
            }
        }
        cout << endl;
    }
    infile.close();
    return 0;
}

File: src/java/spiral_printing.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

/**
 * Solution for http://codeeval.com/open_challenges/57/
 */
public class spiral_printing {

    public static void printSpiral(String[][] matrix, final PrintWriter out) {
        while (!isEmpty(matrix)) {
            printFirstRow(matrix, out);
            matrix = removeFirstRow(matrix);
            if (!isEmpty(matrix)) {
                out.print(' ');
                matrix = rotateCounterClockWise(matrix);
            }
        }
    }

    private static boolean isEmpty(final String[][] matrix) {
        return matrix.length == 0;
    }

    private static void printFirstRow(final String[][] matrix, final PrintWriter out) {
        final String[] firstRow = matrix[0];
        for (int i = 0; i < firstRow.length; i++) {
            out.print(firstRow[i]);
            if (i != firstRow.length - 1) {
                out.print(' ');
            }
        }
    }

    private static String[][] removeFirstRow(final String[][] matrix) {
        final int rows = matrix.length;
        final int cols = matrix[0].length;
        final String[][] newMatrix = new String[rows - 1][cols];
        System.arraycopy(matrix, 1, newMatrix, 0, rows - 1);
        return newMatrix;
    }

    private static String[][] rotateCounterClockWise(final String[][] matrix) {
        final int rows = matrix.length;
        final int cols = matrix[0].length;
        final String[][] rotated = new String[cols][rows];
        for (int i = 0; i < rows; i++) {
            for (int j = cols - 1; j >= 0; j--) {
                rotated[cols - 1 - j][i] = matrix[i][j];
            }
        }
        return rotated;
    }

    private static String[][] parseMatrix(final String s) {
        final String[] tokens = s.split(";");
        final int rows = Integer.valueOf(tokens[0]);
        final int cols = Integer.valueOf(tokens[1]);
        final String[][] matrix = new String[rows][cols];
        final Scanner scanner = new Scanner(tokens[2]);
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = scanner.next();
            }
        }
        return matrix;
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.err.println("usage: java " + spiral_printing.class.getName() + " filename");
            System.exit(1);
        }
        BufferedReader in = null;
        PrintWriter out = null;
        try {
            in = new BufferedReader(new FileReader(args[0]));
            out = new PrintWriter(System.out);
            String line;
            while ((line = in.readLine()) != null) {
                printSpiral(parseMatrix(line), out);
                out.println();
            }
        } finally {
            if (in != null) {
                in.close();
            }
            if (out != null) {
                out.close();
            }
        }
    }
}
